File: web/src/store/party.ts
import { create } from "zustand";
import type { PartyMember } from "@/types";

interface PartyStoreState {
  code: string | null;
  active: boolean;
  name: string;
  isHost: boolean;
  currentIndex: number;
  members: PartyMember[];

  // Host-authoritative playback state (mirrored from the SSE stream). Guests
  // follow these; the host owns them.
  isPlaying: boolean;
  positionSec: number;
  /** epoch ms of the last playback update, or null if never set. */
  playbackUpdatedAt: number | null;

  setParty: (state: {
    code: string;
    active?: boolean;
    name: string;
    isHost: boolean;
    currentIndex: number;
    members: PartyMember[];
    isPlaying: boolean;
    positionSec: number;
    playbackUpdatedAt: number | null;
  }) => void;
  clearParty: () => void;
}

export const usePartyStore = create<PartyStoreState>((set) => ({
  code: null,
  active: false,
  name: "",
  isHost: false,
  currentIndex: -1,
  members: [],
  isPlaying: false,
  positionSec: 0,
  playbackUpdatedAt: null,

  setParty: (state) =>
    set({
      code: state.code,
      active: state.active ?? true,
      name: state.name,
      isHost: state.isHost,
      currentIndex: state.currentIndex,
      members: state.members,
      isPlaying: state.isPlaying,
      positionSec: state.positionSec,
      playbackUpdatedAt: state.playbackUpdatedAt,
    }),

  clearParty: () =>
    set({
      code: null,
      active: false,
      name: "",
      isHost: false,
      currentIndex: -1,
      members: [],
      isPlaying: false,
      positionSec: 0,
      playbackUpdatedAt: null,
    }),
}));

File: api/app/schemas/party.py
"""Schemas for collaborative party sessions."""
from pydantic import BaseModel, Field

from .track import ArtistRef


class PartyTrackOut(BaseModel):
    id: int
    deezer_id: str
    title: str = ""
    artist: str = ""
    artist_id: str = ""
    artists: list[ArtistRef] = Field(default_factory=list)
    album: str = ""
    album_id: str = ""
    cover: str = ""
    duration_sec: int = 0
    added_by: str = ""


class PartyMemberOut(BaseModel):
    name: str = ""
    avatar_url: str | None = None


class PartyState(BaseModel):
    code: str
    name: str = ""
    host_name: str = ""
    is_host: bool = False
    current_index: int = -1
    # Host-authoritative playback state, broadcast to guests.
    is_playing: bool = False
    position_sec: float = 0.0
    # ISO-8601 timestamp of the last playback update, or None if never set.
    playback_updated_at: str | None = None
    members: list[PartyMemberOut] = Field(default_factory=list)
    tracks: list[PartyTrackOut] = Field(default_factory=list)
